lib/puppet/provider/foreman_resource/rest_v3.rb

```diff
@@ -95,6 +98,21 @@ def success?(response)
   end
 
   def error_message(response)
-    JSON.parse(response.body)['error']['full_messages'].join(' ') rescue "unknown error (response #{response.code})"
+    messages = {
+      '400' => '400 Bad Request: Something is wrong with the data we sent to Foreman server',
+      '401' => '401 Unauthorized Request: Check credentials for validity',
+      '403' => '403 Forbidden: The credentials were valid but do not grant access to the requested resource',
+      '404' => '404 Not Found: The requested resource was not found',
+      '500' => '500 Internal Server Error: Check /var/log/foreman/production.log on Foreman server for detailed information',
+      '502' => '502 Bad Gateway: The webserver received an invalid response from the application server. Was Foreman unable to handle the request?',
+      '503' => '503 Service Unavailable: The webserver was unable to reach the backend service. Is foreman.service running?',
+      '504' => '504 Gateway Timeout: The webserver timed out waiting for a response from the application. Is Foreman under unusually heavy load?'
```

Comment:
I wonder if we could say `Apache` instead of `webserver` here. If I am user seeing these, they sound generic enough that I would have to think about what the webserver vs the application was.

Reply:
I think branding would be extremely helpful here. I can see at least 3 uses that substantively improve this feature:

1. Being properly specific with regards to 'SmartProxy' vs. 'Capsule' and similar brand specific terms
2. Linking to relevant KCS in downstream branded error messages
3. Being more specific in message language ('Apache' vs. 'Webserver') due to less variety in deployments

Reply:
For the 504, as an example, could it read "504 Gateway Timeout: The webserver timed out waiting for a response from Foreman" or `foreman.service` as you used in the 503.

I think 'Smart Proxy' is the right term regardless of branding as that is the software. Though, arguably, its 'foreman-proxy' and thus rightfully confusing :)

Reply:
In vanilla Foreman you see `Smart Proxies` in the menu so there it's all consistent. The API endpoint is also `/smart_proxies`.

Arguably you could say that Foreman Proxy is just the default implementation but in theory any service could implement it. For example, a DHCP appliance can implement it and be registered directly. For Pulp you can also write a plugin `pulpcore_smart_proxy` to replace `smart_proxy_pulp`.